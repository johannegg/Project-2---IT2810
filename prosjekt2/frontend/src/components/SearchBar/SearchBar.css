.searchContainer {
	width: 30%;
	position: absolute;
	left: auto;
	margin-bottom: 20px;
	top: 160px;
}

.searchInput {
	width: 100%;
	border: 3px solid var(--theme-green);
	padding: 5px;
	padding-left: 35px;
	height: 25px;
	border-radius: 20px;
	outline: none;
	background-color: white;
}

.searchInput::placeholder {
	color: var(--theme-darker-green);
}

.searchInput:focus{
	border: 3px solid var(--theme-dark-green);
}

.searchContainer:focus-within .searchIcon {
	color: var(--theme-dark-green);
}

.searchIconContainer {
	width: 37px;
	padding: 0;
	height: 90%;
	font-size: 20px;
	position: relative;
	right: 45px;
	top: 9px;
	background-color: transparent;
}

.searchButton:focus {
	outline: none;
}

.searchIcon {
	font-size: 19px;
	color: var(--theme-green);
}

.iconContainer {
	position: absolute;
	left: 12px;
	top: 11px;
}

@media (max-width: 700px) {
	.searchContainer {
		left: 18%;
		width: 54%;
	}
}
