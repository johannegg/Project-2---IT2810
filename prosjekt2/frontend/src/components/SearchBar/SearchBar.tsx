import React, { useEffect, useState } from "react";
import { FaSearch } from "react-icons/fa";
import "./SearchBar.css";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faXmark } from "@fortawesome/free-solid-svg-icons";

type SearchBarProps = {
	setSearchTerm: (term: string) => void;
	initialSearchTerm?: string; // Nytt prop for å sette en initial verdi
};

export const SearchBar: React.FC<SearchBarProps> = ({ setSearchTerm, initialSearchTerm = "" }) => {
	const [searchInput, setSearchInput] = useState<string>(initialSearchTerm);

	useEffect(() => {
		setSearchInput(initialSearchTerm); // Sett initial verdi ved første render
	}, [initialSearchTerm]);

	const handleSearchSubmit = (event: React.FormEvent<HTMLFormElement>) => {
		event.preventDefault();
		setSearchTerm(searchInput);
	};

	const clearInput = () => {
		setSearchInput("");
		setSearchTerm("");
	};

	return (
		<div className="searchContainer" aria-label="Search bar container">
			<form className="searchForm" onSubmit={handleSearchSubmit} aria-label="Search Form">
				<input
					className="searchInput"
					placeholder="Search for a song or an artist"
					value={searchInput}
					onChange={(e) => setSearchInput(e.target.value)}
					aria-label="Search Input Field"
				/>
				<button type="submit" className="iconContainer" aria-label="Submit Search">
					<FaSearch className="searchIcon" aria-hidden="true" />
				</button>
			</form>
			{searchInput && (
				<button
					className="clearButton"
					onClick={clearInput}
					aria-label="Clear Search Button"
				>
					<FontAwesomeIcon icon={faXmark} aria-hidden="true" />
				</button>
			)}
		</div>
	);
};
