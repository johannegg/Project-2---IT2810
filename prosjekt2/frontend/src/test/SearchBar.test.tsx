import { describe, test, vi, beforeEach, expect } from "vitest";
import { render, screen, fireEvent } from "@testing-library/react";
import { SearchBar } from "../components/SearchBar/SearchBar";

describe("SearchBar Component", () => {
    let setSearchTermMock: ReturnType<typeof vi.fn>;

    beforeEach(() => {
        setSearchTermMock = vi.fn();
    });

    test("renders the SearchBar component", () => {
        const { asFragment } = render(<SearchBar setSearchTerm={setSearchTermMock} />);
        
        // Verify the input field and buttons are present
        const input = screen.getByLabelText(/search input field/i);
        expect(input).toBeInTheDocument();

        const searchButton = screen.getByLabelText(/submit search/i);
        expect(searchButton).toBeInTheDocument();

        // Snapshot the initial render
        expect(asFragment()).toMatchSnapshot();
    });

    test("accepts an initial search term", () => {
        const { asFragment } = render(<SearchBar setSearchTerm={setSearchTermMock} initialSearchTerm="Initial Value" />);
    
        // Verify the input field has the initial value
        const input = screen.getByLabelText(/search input field/i);
        expect(input).toHaveValue("Initial Value");

        // Snapshot the render with the initial value
        expect(asFragment()).toMatchSnapshot();
    });

    test("updates the input value as the user types", () => {
        const { asFragment } = render(<SearchBar setSearchTerm={setSearchTermMock} />);

        // Simulate typing in the input field
        const input = screen.getByLabelText(/search input field/i);
        fireEvent.change(input, { target: { value: "New Value" } });

        // Verify the input value has been updated
        expect(input).toHaveValue("New Value");

        // Snapshot the render after typing
        expect(asFragment()).toMatchSnapshot();
    });

    test("calls setSearchTerm when the form is submitted", () => {
        const { asFragment } = render(<SearchBar setSearchTerm={setSearchTermMock} />);

        // Simulate typing and form submission
        const input = screen.getByLabelText(/search input field/i);
        fireEvent.change(input, { target: { value: "Search Term" } });

        const form = screen.getByLabelText(/search form/i);
        fireEvent.submit(form);

        // Verify setSearchTerm is called with the correct value
        expect(setSearchTermMock).toHaveBeenCalledWith("Search Term");

        // Snapshot the render after submission
        expect(asFragment()).toMatchSnapshot();
    });

    test("clears the input when the clear button is clicked", () => {
        const { asFragment } = render(<SearchBar setSearchTerm={setSearchTermMock} initialSearchTerm="Clear Me" />);
    
        // Verify the clear button is present
        const clearButton = screen.getByLabelText(/clear search button/i);
        expect(clearButton).toBeInTheDocument();
    
        // Simulate clicking the clear button
        fireEvent.click(clearButton);
    
        // Verify the input is cleared
        const input = screen.getByLabelText(/search input field/i);
        expect(input).toHaveValue("");
    
        // Verify setSearchTerm is called with an empty string
        expect(setSearchTermMock).toHaveBeenCalledWith("");

        // Snapshot the render after clearing
        expect(asFragment()).toMatchSnapshot();
    });
});
