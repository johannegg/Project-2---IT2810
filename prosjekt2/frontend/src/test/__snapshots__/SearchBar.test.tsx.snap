// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchBar Component > accepts an initial search term 1`] = `
<DocumentFragment>
  <div
    aria-label="Search bar container"
    class="searchContainer"
  >
    <form
      aria-label="Search Form"
      class="searchForm"
    >
      <input
        aria-label="Search Input Field"
        class="searchInput"
        placeholder="Search for a song or an artist"
        value="Initial Value"
      />
      <button
        aria-label="Submit Search"
        class="iconContainer"
        type="submit"
      >
        <svg
          aria-hidden="true"
          class="searchIcon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
          />
        </svg>
      </button>
    </form>
    <button
      aria-label="Clear Search Button"
      class="clearButton"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-xmark "
        data-icon="xmark"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 384 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;

exports[`SearchBar Component > calls setSearchTerm when the form is submitted 1`] = `
<DocumentFragment>
  <div
    aria-label="Search bar container"
    class="searchContainer"
  >
    <form
      aria-label="Search Form"
      class="searchForm"
    >
      <input
        aria-label="Search Input Field"
        class="searchInput"
        placeholder="Search for a song or an artist"
        value="Search Term"
      />
      <button
        aria-label="Submit Search"
        class="iconContainer"
        type="submit"
      >
        <svg
          aria-hidden="true"
          class="searchIcon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
          />
        </svg>
      </button>
    </form>
    <button
      aria-label="Clear Search Button"
      class="clearButton"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-xmark "
        data-icon="xmark"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 384 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;

exports[`SearchBar Component > clears the input when the clear button is clicked 1`] = `
<DocumentFragment>
  <div
    aria-label="Search bar container"
    class="searchContainer"
  >
    <form
      aria-label="Search Form"
      class="searchForm"
    >
      <input
        aria-label="Search Input Field"
        class="searchInput"
        placeholder="Search for a song or an artist"
        value=""
      />
      <button
        aria-label="Submit Search"
        class="iconContainer"
        type="submit"
      >
        <svg
          aria-hidden="true"
          class="searchIcon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
          />
        </svg>
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`SearchBar Component > renders the SearchBar component 1`] = `
<DocumentFragment>
  <div
    aria-label="Search bar container"
    class="searchContainer"
  >
    <form
      aria-label="Search Form"
      class="searchForm"
    >
      <input
        aria-label="Search Input Field"
        class="searchInput"
        placeholder="Search for a song or an artist"
        value=""
      />
      <button
        aria-label="Submit Search"
        class="iconContainer"
        type="submit"
      >
        <svg
          aria-hidden="true"
          class="searchIcon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
          />
        </svg>
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`SearchBar Component > updates the input value as the user types 1`] = `
<DocumentFragment>
  <div
    aria-label="Search bar container"
    class="searchContainer"
  >
    <form
      aria-label="Search Form"
      class="searchForm"
    >
      <input
        aria-label="Search Input Field"
        class="searchInput"
        placeholder="Search for a song or an artist"
        value="New Value"
      />
      <button
        aria-label="Submit Search"
        class="iconContainer"
        type="submit"
      >
        <svg
          aria-hidden="true"
          class="searchIcon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
          />
        </svg>
      </button>
    </form>
    <button
      aria-label="Clear Search Button"
      class="clearButton"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-xmark "
        data-icon="xmark"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 384 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
